go-controller/pkg/crd/adminpolicybasedroute/v1/types.go

```diff
@@ -0,0 +1,125 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:path=adminpolicybasedexternalroute,scope=Cluster
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+
+// AdminPolicyBasedExternalRoute is the Schema for the AdminPolicyBasedExternalRoutes API
+type AdminPolicyBasedExternalRoute struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              AdminPolicyBasedExternalRouteSpec `json:"spec,omitempty"`
+	Status            AdminPolicyBasedRouteStatus       `json:"status,omitempty"`
+}
+
+// AdminPolicyBasedExternalRouteSpec defines the desired state of AdminPolicyBasedExternalRoute
+type AdminPolicyBasedExternalRouteSpec struct {
+	Policies []*ExternalPolicy `json:"policies"`
+}
+
+type ExternalPolicy struct {
+	From     ExternalNetworkSource `json:"from"`
+	NextHops ExternalNextHops      `json:"nextHops"`
+}
+
+type ExternalNetworkSource struct {
+	NamespaceSelector metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+}
+
+type ExternalNextHops struct {
+	// +kubebuilder:validation:UniqueItems=true
+	StaticHops []*StaticHop `json:"static,omitempty"`
+	// +kubebuilder:validation:UniqueItems=true
+	DynamicHops []*DynamicHop `json:"dynamic,omitempty"`
+}
+
+type StaticHop struct {
+	IP           string `json:"ip,omitempty"`
```

Comment:
It would read as duplicated since the `IP` field is part of the `StaticHop` structure:
Example for demonstration purposes:
```go
var instance v1.AdminPolicyBasedExternalRoute
instance.Spec.NextHops.StaticHops[0].IP="127.0.0.1"
```
vs
```go
var instance v1.AdminPolicyBasedExternalRoute
instance.Spec.NextHops.StaticHops[0].StaticHopIP="127.0.0.1"
```